Direct users towards using Rust target feature names in CLI

nagisa opened the pull request:
This PR consists of a couple of changes on how we handle target features.

In particular there is a bug-fix wherein we avoid passing through features that aren't prefixed by `+` or `-` to LLVM. These appear to be causing LLVM to assert, which is pretty poor a behaviour (and also makes it pretty clear we expect feature names to be prefixed).

The other commit, I anticipate to be somewhat more controversial is outputting a warning when users specify a LLVM-specific, or otherwise unknown, feature name on the CLI. In those situations we request users to either replace it with a known Rust feature name (e.g. `bmi` -> `bmi1`) or file a feature request. I've a couple motivations for this: first of all, if users are specifying these features on the command line, I'm pretty confident there is also a need for these features to be usable via `#[cfg(target_feature)]` machinery.  And second, we're growing a fair number of backends recently and having ability to provide some sort of unified-ish interface in this place seems pretty useful to me.

Sponsored by: standard.ai

rust-highfive commented:
r? @estebank

(rust-highfive has picked a reviewer for you, use r? to override)

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -370,36 +366,61 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
+        sess.target
+            .features
+            .split(',')
+            .filter_map(|s| backend_feature_name(s).map(|_| String::from(s))),
```
Likely want to make it a `bug!()` if we get a `None` in this `filter_map`.

bors commented:
:umbrella: The latest upstream changes (presumably #87449) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nagisa commented:
Hm, one remaining problem is that we call `llvm_global_features` for each function so there will be warnings repeated many times for a typical compilation. Will think about how to best address that.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3701525463) [(plain)](https://github.com/rust-lang/rust/commit/4b66ad5cb78275f2eec2258796f6794801656a4a/checks/3701525463/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 12190 tests
.................................................................................................... 100/12190
..............................................ii.......F.......ii..............F.................... 200/12190
.................................................................................................... 400/12190
.................................................................................................... 500/12190
.................................................................................................... 600/12190
..........................................................i......................................... 700/12190
---
failures:

---- [ui] ui/asm/bad-reg.rs stdout ----

error: /checkout/src/test/ui/asm/bad-reg.rs:62: unexpected error: '62:34: 62:48: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/bad-reg.rs:64: unexpected error: '64:34: 64:49: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/bad-reg.rs:66: unexpected error: '66:34: 66:53: register class `ymm_reg` requires the `avx` target feature'
error: 3 unexpected errors found, 0 expected errors not found
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/asm/bad-reg.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/bad-reg" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-C" "target-feature=+avx2" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/bad-reg/auxiliary"
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
    Error {
        line_num: 62,
        kind: Some(
            Error,
            Error,
        ),
        msg: "62:34: 62:48: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 64,
        kind: Some(
            Error,
            Error,
        ),
        msg: "64:34: 64:49: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 66,
        kind: Some(
            Error,
            Error,
        ),
        msg: "66:34: 66:53: register class `ymm_reg` requires the `avx` target feature",
]

thread '[ui] ui/asm/bad-reg.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:1532:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [ui] ui/asm/type-check-3.rs stdout ----

error: /checkout/src/test/ui/asm/type-check-3.rs:25: unexpected error: '25:22: 25:50: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/type-check-3.rs:26: unexpected error: '26:20: 26:33: register class `kreg` requires at least one target feature: avx512bw, avx512f'

error: /checkout/src/test/ui/asm/type-check-3.rs:27: unexpected error: '27:20: 27:33: register class `kreg` requires at least one target feature: avx512bw, avx512f'

error: /checkout/src/test/ui/asm/type-check-3.rs:39: unexpected error: '39:20: 39:36: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/type-check-3.rs:41: unexpected error: '41:20: 41:51: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/type-check-3.rs:45: unexpected error: '45:22: 45:38: register class `ymm_reg` requires the `avx` target feature'

error: /checkout/src/test/ui/asm/type-check-3.rs:46: unexpected error: '46:22: 46:53: register class `ymm_reg` requires the `avx` target feature'
error: /checkout/src/test/ui/asm/type-check-3.rs:12: expected error not found: type `i128` cannot be used with this register class


error: /checkout/src/test/ui/asm/type-check-3.rs:14: expected error not found: type `__m128` cannot be used with this register class

error: /checkout/src/test/ui/asm/type-check-3.rs:16: expected error not found: type `__m256` cannot be used with this register class
error: /checkout/src/test/ui/asm/type-check-3.rs:18: expected error not found: type `u8` cannot be used with this register class


error: /checkout/src/test/ui/asm/type-check-3.rs:27: expected error not found: `avx512bw` target feature is not enabled
error: /checkout/src/test/ui/asm/type-check-3.rs:32: expected warning not found: formatting may not be suitable for sub-register argument

error: /checkout/src/test/ui/asm/type-check-3.rs:34: expected warning not found: formatting may not be suitable for sub-register argument


error: /checkout/src/test/ui/asm/type-check-3.rs:36: expected warning not found: formatting may not be suitable for sub-register argument

error: /checkout/src/test/ui/asm/type-check-3.rs:39: expected warning not found: formatting may not be suitable for sub-register argument

error: /checkout/src/test/ui/asm/type-check-3.rs:50: expected error not found: type `i8` cannot be used with this register class

error: /checkout/src/test/ui/asm/type-check-3.rs:62: expected error not found: incompatible types for asm inout argument

error: /checkout/src/test/ui/asm/type-check-3.rs:64: expected error not found: incompatible types for asm inout argument

error: /checkout/src/test/ui/asm/type-check-3.rs:66: expected error not found: incompatible types for asm inout argument
error: /checkout/src/test/ui/asm/type-check-3.rs:82: expected error not found: constants cannot refer to statics

error: /checkout/src/test/ui/asm/type-check-3.rs:85: expected error not found: constants cannot refer to statics


error: /checkout/src/test/ui/asm/type-check-3.rs:88: expected error not found: constants cannot refer to statics

error: 7 unexpected errors found, 16 expected errors not found
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/asm/type-check-3.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/type-check-3" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-C" "target-feature=+avx512f" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/type-check-3/auxiliary"
    Error {
        line_num: 25,
        kind: Some(
            Error,
            Error,
        ),
        msg: "25:22: 25:50: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 26,
        kind: Some(
            Error,
            Error,
        ),
        msg: "26:20: 26:33: register class `kreg` requires at least one target feature: avx512bw, avx512f",
    Error {
        line_num: 27,
        kind: Some(
            Error,
            Error,
        ),
        msg: "27:20: 27:33: register class `kreg` requires at least one target feature: avx512bw, avx512f",
    Error {
        line_num: 39,
        kind: Some(
            Error,
            Error,
        ),
        msg: "39:20: 39:36: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 41,
        kind: Some(
            Error,
            Error,
        ),
        msg: "41:20: 41:51: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 45,
        kind: Some(
            Error,
            Error,
        ),
        msg: "45:22: 45:38: register class `ymm_reg` requires the `avx` target feature",
    Error {
        line_num: 46,
        kind: Some(
            Error,
            Error,
        ),
        msg: "46:22: 46:53: register class `ymm_reg` requires the `avx` target feature",
]

not found errors (from test file): [
    Error {
---
        line_num: 14,
        kind: Some(
            Error,
        ),
        msg: "type `__m128` cannot be used with this register class",
    Error {
        line_num: 16,
        kind: Some(
            Error,
            Error,
        ),
        msg: "type `__m256` cannot be used with this register class",
    Error {
        line_num: 18,
        kind: Some(
            Error,
---
        line_num: 27,
        kind: Some(
            Error,
        ),
        msg: "`avx512bw` target feature is not enabled",
    Error {
        line_num: 32,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "formatting may not be suitable for sub-register argument",
    Error {
        line_num: 34,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "formatting may not be suitable for sub-register argument",
    Error {
        line_num: 36,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "formatting may not be suitable for sub-register argument",
    Error {
        line_num: 39,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "formatting may not be suitable for sub-register argument",
    Error {
        line_num: 50,
        kind: Some(
            Error,
---
        line_num: 62,
        kind: Some(
            Error,
        ),
        msg: "incompatible types for asm inout argument",
    Error {
        line_num: 64,
        kind: Some(
            Error,
            Error,
        ),
        msg: "incompatible types for asm inout argument",
    Error {
        line_num: 66,
        kind: Some(
            Error,
            Error,
        ),
        msg: "incompatible types for asm inout argument",
    Error {
        line_num: 82,
        kind: Some(
            Error,
---
test result: FAILED. 12086 passed; 2 failed; 102 ignored; 0 measured; 0 filtered out; finished in 127.42s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:12:20
```

</details>

nagisa commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit e25afb20b11f9cc764ea5e1f6d593b504bc1807f with merge 0c190043f011023658fc263567eaf5bc76f3cb4b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e25afb20b11f9cc764ea5e1f6d593b504bc1807f","merge_sha":"0c190043f011023658fc263567eaf5bc76f3cb4b"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3702356679) [(plain)](https://github.com/rust-lang/rust/commit/e25afb20b11f9cc764ea5e1f6d593b504bc1807f/checks/3702356679/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_llvm/src/lib.rs at line 210:
     }
 
     fn provide(&self, providers: &mut Providers) {
-        providers.global_backend_features = |tcx, ()| {
-            llvm_util::global_llvm_features(tcx.sess)
-        }
+        providers.global_backend_features = |tcx, ()| llvm_util::global_llvm_features(tcx.sess)
 
 
     fn print(&self, req: PrintRequest, sess: &Session) {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_llvm/src/lib.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/diagnostic.rs" "/checkout/compiler/rustc_codegen_llvm/src/consts.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/archive_ro.rs" "/checkout/compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs" "/checkout/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/lto.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3703179669)
Build commit: 0c190043f011023658fc263567eaf5bc76f3cb4b (`0c190043f011023658fc263567eaf5bc76f3cb4b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3703179669"},"merge_sha":"0c190043f011023658fc263567eaf5bc76f3cb4b"} -->

rust-timer commented:
Queued 0c190043f011023658fc263567eaf5bc76f3cb4b with parent a0648eab3686f100c7ab9b0d38472c740079cce4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a0648eab3686f100c7ab9b0d38472c740079cce4&end=0c190043f011023658fc263567eaf5bc76f3cb4b).

rust-timer commented:
Finished benchmarking commit (0c190043f011023658fc263567eaf5bc76f3cb4b): [comparison url](https://perf.rust-lang.org/compare.html?start=a0648eab3686f100c7ab9b0d38472c740079cce4&end=0c190043f011023658fc263567eaf5bc76f3cb4b).

**Summary**: This change led to very large relevant regressions 😿 in compiler performance.
- Very large regression in instruction counts (up to 4651.6% on `incr-unchanged` builds of `webrender`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nagisa commented:
4 digits of regression. Nice.

nagisa commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 69025a6734909ab8b62f732881c9c9133b0ac082 with merge fb098907a7ca48eb029cdb8ef9542769ebfc5238...
<!-- homu: {"type":"TryBuildStarted","head_sha":"69025a6734909ab8b62f732881c9c9133b0ac082","merge_sha":"fb098907a7ca48eb029cdb8ef9542769ebfc5238"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3705177942)
Build commit: fb098907a7ca48eb029cdb8ef9542769ebfc5238 (`fb098907a7ca48eb029cdb8ef9542769ebfc5238`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3705177942"},"merge_sha":"fb098907a7ca48eb029cdb8ef9542769ebfc5238"} -->

rust-timer commented:
Queued fb098907a7ca48eb029cdb8ef9542769ebfc5238 with parent f06f9bbd3a2b0a2781decd6163b14f71dd59bf7f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f06f9bbd3a2b0a2781decd6163b14f71dd59bf7f&end=fb098907a7ca48eb029cdb8ef9542769ebfc5238).

rust-timer commented:
Finished benchmarking commit (fb098907a7ca48eb029cdb8ef9542769ebfc5238): [comparison url](https://perf.rust-lang.org/compare.html?start=f06f9bbd3a2b0a2781decd6163b14f71dd59bf7f&end=fb098907a7ca48eb029cdb8ef9542769ebfc5238).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

bjorn3 commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
+                            s
+                        ));
+                        diag.note("features must begin with a `+` to enable or `-` to disable it");
+                        diag.emit();
```
I think this warning should be moved to the driver, so it also works for other codegen backends and for `cargo check`.

nagisa replied:
Given that each backend is currently in charge of parsing everything about the target features, except for splitting by comma. I don't think right now is the right time to move this diagnostic to anywhere.

I do have more changes I want to make with regards to how we handle target features, and moving some of that responsibility away from `rustc_codegen_*` is one of those things. As an example, the only thing codegen backends should probably know about “rustc” target features is how to convert a rustc target feature to its own internal interpretation.

But I think some more refactors are in order before this can happen.

estebank commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
+                        "unknown feature specified for `-Ctarget-feature`: `{}`",
+                        feature
+                    ));
+                    if let Some(rust_feature) = rust_feature {
+                        diag.note("it is still passed through to the codegen backend");
+                        diag.help(&format!("did you mean: `{}`", rust_feature));
+                    } else {
+                        diag.note("it is still passed through to the codegen backend");
+                        diag.note("consider filing a feature request");
+                    }
+                    diag.emit();
```
```suggestion
                    diag.note("it is still passed through to the codegen backend");
                    if let Some(rust_feature) = rust_feature {
                        diag.help(&format!("you might have meant: `{}`", rust_feature));
                    } else {
                        diag.note("consider filing a feature request");
                    }
                    diag.emit();
```

Not too happy with the "consider filing a feature request", but I have no better idea :-/

nagisa replied:
Hm, is there a reason why this wording makes you concerned? My goal is basically to drive people to describe their use-cases for specific features that they are using so that we could add them to `rustc`.

estebank replied:
It gives the user something to do, but it is kind of... vague. I'm ok with landing it as is though.

estebank commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
+                // Warn against use of LLVM specific feature names on the CLI.
+                if !supported_features.iter().any(|&(v, _)| v == feature) {
+                    let rust_feature = supported_features.iter().find_map(|&(rust_feature, _)| {
+                        let llvm_features = to_llvm_features(sess, rust_feature);
```
Should we be doing fuzzy searching in `to_llvm_features` for the suggestion given bellow?

nagisa replied:
Possibly. My primary goal was to direct users from using LLVM names towards using Rust names, and much less so to handle typos better. I do agree however that a fuzzy suggestions here would be ideal. Even better if we make them work for more than just `-Ctarget-features`.

estebank reviewed (approved):
The code changes look reasonable.

bors commented:
:umbrella: The latest upstream changes (presumably #89183) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pnkfelix commented:
AFAICT this has been r+'ed and just needs a rebate. Changing to S-waiting-on-author.

JohnCSimon commented:
> AFAICT this has been r+'ed and just needs a rebase. Changing to S-waiting-on-author.

@nagisa ☝️

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5259239853?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/19b61538febdb83d0a644b9b98a149c9f9d93997/checks/5259239853/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking ar v0.8.0
    Checking gccjit_sys v0.0.1 (https://github.com/antoyo/gccjit.rs#0672b78d)
    Checking gccjit v1.0.0 (https://github.com/antoyo/gccjit.rs#0672b78d)
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0050]: method `target_machine_factory` has 3 parameters but the declaration in trait `target_machine_factory` has 4
    |
    |
135 |     fn target_machine_factory(&self, _sess: &Session, _opt_level: OptLevel) -> TargetMachineFactoryFn<Self> {
    |
    |
    = note: `target_machine_factory` from trait: `fn(&Self, &Session, OptLevel, &[std::string::String]) -> Arc<(dyn std::ops::Fn(TargetMachineFactoryConfig) -> Result<<Self as WriteBackendMethods>::TargetMachine, std::string::String> + Send + std::marker::Sync + 'static)>`
For more information about this error, try `rustc --explain E0050`.
error: could not compile `rustc_codegen_gcc` due to previous error
Build completed unsuccessfully in 0:03:27
```

</details>

nagisa commented:
Finally got around to rebasing this. The surrounding code changed quite a bit since since July so a re-review is probably warranted. I didn't really make any major functional changes since the last review though, and so all comments/responses above stand.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5259306911?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/52f7529a24bb96f131f773c8cc4977ebd4409ad5/checks/5259306911/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking ar v0.8.0
    Checking gccjit_sys v0.0.1 (https://github.com/antoyo/gccjit.rs#0672b78d)
    Checking gccjit v1.0.0 (https://github.com/antoyo/gccjit.rs#0672b78d)
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0050]: method `target_machine_factory` has 3 parameters but the declaration in trait `target_machine_factory` has 4
    |
    |
135 |     fn target_machine_factory(&self, _sess: &Session, _opt_level: OptLevel) -> TargetMachineFactoryFn<Self> {
    |
    |
    = note: `target_machine_factory` from trait: `fn(&Self, &Session, OptLevel, &[std::string::String]) -> Arc<(dyn std::ops::Fn(TargetMachineFactoryConfig) -> Result<<Self as WriteBackendMethods>::TargetMachine, std::string::String> + Send + std::marker::Sync + 'static)>`
For more information about this error, try `rustc --explain E0050`.
error: could not compile `rustc_codegen_gcc` due to previous error
Build completed unsuccessfully in 0:03:48
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5259386697?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b693b2afd8d52edd4073ff3c7446f5f8d6c181c4/checks/5259386697/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error: unused variable: `features`
   --> src/lib.rs:135:77
    |
135 |     fn target_machine_factory(&self, _sess: &Session, _opt_level: OptLevel, features: &[String]) -> TargetMachineFactoryFn<Self> {
    |                                                                             ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_features`
    |
    = note: `-D unused-variables` implied by `-D warnings`
error: could not compile `rustc_codegen_gcc` due to previous error
Build completed unsuccessfully in 0:03:37
```

</details>

estebank commented on src/test/ui/target-feature/missing-plusminus.stderr:
```diff
@@ -0,0 +1,6 @@
+warning: unknown feature specified for `-Ctarget-feature`: `banana`
+   |
+   = note: features must begin with a `+` to enable or `-` to disable it
```
Should we also have a test for `-Ctarget-feature=rdrand` where we tell them "you have to either enable it or disable it"?

nagisa replied:
Done!

nagisa commented:
AFAIK beta was cut on Friday already, but I don't anticipate coming back to take a look at your review comment before the weekend anyway, so this is definitely not landing before train rollover.

bors commented:
:umbrella: The latest upstream changes (presumably #94221) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nagisa commented:
@bors r=estebank

bors commented:
:pushpin: Commit 0c0081ab99aede50d08c93f93134c80d74670a9e has been approved by `estebank`

<!-- @bors r=estebank 0c0081ab99aede50d08c93f93134c80d74670a9e -->
<!-- homu: {"type":"Approved","sha":"0c0081ab99aede50d08c93f93134c80d74670a9e","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit 0c0081ab99aede50d08c93f93134c80d74670a9e with merge 782c13b3b2d9f2fec2a7fb774937d53ac01f2c6f...
<!-- homu: {"type":"BuildStarted","head_sha":"0c0081ab99aede50d08c93f93134c80d74670a9e","merge_sha":"782c13b3b2d9f2fec2a7fb774937d53ac01f2c6f"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5351577280?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5351577280?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`i686-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5351489264?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/782c13b3b2d9f2fec2a7fb774937d53ac01f2c6f/checks/5351489264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] ui/target-feature/tied-features.rs ... ignored
test [ui] ui/target-feature/wasm-safe.rs ... ignored
test [ui] ui/symbol-names/trait-objects.rs#v0 ... ok
test [ui] ui/target-feature/invalid-attribute.rs ... ok
test [ui] ui/target-feature/missing-plusminus.rs ... ok
test [ui] ui/target-feature/similar-feature-suggestion.rs ... ok
test [ui] ui/target-feature/missing-plusminus-2.rs ... FAILED
test [ui] ui/target-feature/unknown-feature.rs ... ok
test [ui] ui/symbol-names/verbose.rs ... ok
test [ui] ui/terminal-width/non-1-width-unicode-multiline-label.rs ... ok
test [ui] ui/terminal-width/tabs-trimming.rs ... ok
test [ui] ui/target-feature/rust-specific-name-no-warnings.rs ... ok
---
test [ui] ui/wrong-hashset-issue-42918.rs ... ok

failures:

---- [ui] ui/target-feature/missing-plusminus-2.rs stdout ----
error: test compilation failed although it shouldn't!
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/i686-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/missing-plusminus-2.rs" "-Zthreads=1" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/i686-unknown-linux-gnu/test/ui/target-feature/missing-plusminus-2" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Ctarget-feature=rdrand" "--crate-type=rlib" "--target=x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/i686-unknown-linux-gnu/test/ui/target-feature/missing-plusminus-2/auxiliary"
stdout: none
--- stderr -------------------------------
   |
   |
   = note: the `x86_64-unknown-linux-gnu` target may not be installed
   = help: consider downloading the target with `rustup target add x86_64-unknown-linux-gnu`
   = help: consider building the standard library from source with `cargo build -Zbuild-std`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
------------------------------------------
------------------------------------------



failures:
    [ui] ui/target-feature/missing-plusminus-2.rs
test result: FAILED. 12496 passed; 1 failed; 172 ignored; 0 measured; 0 filtered out; finished in 127.35s

Some tests failed in compiletest suite=ui mode=ui host=i686-unknown-linux-gnu target=i686-unknown-linux-gnu
Build completed unsuccessfully in 0:18:02
```

</details>

nagisa commented:
@bors r=estebank

bors commented:
:pushpin: Commit d3a9c31c302683d4a65626f23f441095bb9ce12f has been approved by `estebank`

<!-- @bors r=estebank d3a9c31c302683d4a65626f23f441095bb9ce12f -->
<!-- homu: {"type":"Approved","sha":"d3a9c31c302683d4a65626f23f441095bb9ce12f","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit d3a9c31c302683d4a65626f23f441095bb9ce12f with merge 0a7ce2be4863511ce820c7308c165e94cb247280...
<!-- homu: {"type":"BuildStarted","head_sha":"d3a9c31c302683d4a65626f23f441095bb9ce12f","merge_sha":"0a7ce2be4863511ce820c7308c165e94cb247280"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5352384447?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5352384447?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5352296278?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/0a7ce2be4863511ce820c7308c165e94cb247280/checks/5352296278/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] ui/target-feature/rust-specific-name-no-warnings.rs ... ignored
test [ui] ui/symbol-names/issue-75326.rs#v0 ... ok
test [ui] ui/tail-typeck.rs ... ok
test [ui] ui/tag-that-dare-not-speak-its-name.rs ... ok
test [ui] ui/target-feature/similar-feature-suggestion.rs ... ok
test [ui] ui/target-feature/missing-plusminus.rs ... ok
test [ui] ui/target-feature/missing-plusminus-2.rs ... ok
test [ui] ui/target-feature/wasm-safe.rs ... ignored
test [ui] ui/target-feature/tied-features-cli.rs#four ... ok
test [ui] ui/tag-variant-cast-non-nullary.rs ... ok
test [ui] ui/target-feature/tied-features-cli.rs#three ... FAILED
---
failures:

---- [ui] ui/target-feature/tied-features-cli.rs#three stdout ----

error in revision `three`: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/tied-features-cli.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--cfg" "three" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.three" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "target-feature=+paca,+pacg,-paca" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.three/auxiliary"
stdout: none
Some tests failed in compiletest suite=ui mode=ui host=aarch64-unknown-linux-gnu target=aarch64-unknown-linux-gnu
stderr: none

---- [ui] ui/target-feature/tied-features-cli.rs#two stdout ----


error in revision `two`: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/tied-features-cli.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--cfg" "two" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.two" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "target-feature=-pacg,+pacg" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.two/auxiliary"
stdout: none
stderr: none

---- [ui] ui/target-feature/tied-features-cli.rs#one stdout ----


error in revision `one`: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/tied-features-cli.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--cfg" "one" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.one" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "target-feature=+paca" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/target-feature/tied-features-cli.one/auxiliary"
stdout: none
stderr: none


failures:
    [ui] ui/target-feature/tied-features-cli.rs#one
```

</details>

nagisa commented:
@bors r=estebank

There was a platform-specific test that broke. I fixed the relevant functionality (which involved mostly just calling into the right function at a different time) and made the tests `no_core` so they are also runnable by developers running other architectures.

bors commented:
:pushpin: Commit 7c696fb98172b92b6079abebc5332699fdf580df has been approved by `estebank`

<!-- @bors r=estebank 7c696fb98172b92b6079abebc5332699fdf580df -->
<!-- homu: {"type":"Approved","sha":"7c696fb98172b92b6079abebc5332699fdf580df","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit 7c696fb98172b92b6079abebc5332699fdf580df with merge 554329d7c4859961c091a13e47838dc5e07dbc67...
<!-- homu: {"type":"BuildStarted","head_sha":"7c696fb98172b92b6079abebc5332699fdf580df","merge_sha":"554329d7c4859961c091a13e47838dc5e07dbc67"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5353523545?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/554329d7c4859961c091a13e47838dc5e07dbc67/checks/5353523545/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] ui/tag-that-dare-not-speak-its-name.rs ... ok
test [ui] ui/symbol-names/issue-60925.rs#v0 ... ok
test [ui] ui/symbol-names/issue-60925.rs#legacy ... ok
test [ui] ui/target-feature/tied-features-cli.rs#one ... ok
test [ui] ui/target-feature/similar-feature-suggestion.rs ... ok
test [ui] ui/target-feature/missing-plusminus.rs ... ok
test [ui] ui/target-feature/tied-features-cli.rs#three ... ok
test [ui] ui/target-feature/tied-features-cli.rs#three ... ok
test [ui] ui/target-feature/missing-plusminus-2.rs ... ok
test [ui] ui/tag-variant-cast-non-nullary.rs ... ok
test [ui] ui/target-feature/unknown-feature.rs ... FAILED
test [ui] ui/target-feature/tied-features.rs ... ok
test [ui] ui/target-feature/wasm-safe.rs ... ok
---
test [ui] ui/wrong-hashset-issue-42918.rs ... ok

failures:

---- [ui] ui/target-feature/unknown-feature.rs stdout ----

6 
6 
7 '+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
8 '+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
+ '+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
+ '+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
10 
11 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/unknown-feature/unknown-feature.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args target-feature/unknown-feature.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/target-feature/unknown-feature.rs" "-Zthreads=1" "--target=wasm32-unknown-unknown" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/unknown-feature" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-unknown-unknown/native/rust-test-helpers" "-Ctarget-feature=+definitely-unknown-feature" "--crate-type=rlib" "-Ccodegen-units=1" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/target-feature/unknown-feature/auxiliary"
stdout: none
--- stderr -------------------------------
'+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
warning: unknown feature specified for `-Ctarget-feature`: `definitely-unknown-feature`
   = note: it is still passed through to the codegen backend
   = note: consider filing a feature request


'+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
'+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
'+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
'+definitely-unknown-feature' is not a recognized feature for this target (ignoring feature)
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5353660921?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5353660921?check_suite_focus=true","builder_name":"checks-actions"} -->

nagisa commented:
The failing test seems like it'll be flaky and it doesn't provide much value either way, so I went ahead and removed it entirely.

@bors r=estebank

bors commented:
:pushpin: Commit df701a292ce552fddad2048cfcb6edaf90222d85 has been approved by `estebank`

<!-- @bors r=estebank df701a292ce552fddad2048cfcb6edaf90222d85 -->
<!-- homu: {"type":"Approved","sha":"df701a292ce552fddad2048cfcb6edaf90222d85","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit df701a292ce552fddad2048cfcb6edaf90222d85 with merge adcccf31a23c7eedff01f672159a3f4b724559b3...
<!-- homu: {"type":"BuildStarted","head_sha":"df701a292ce552fddad2048cfcb6edaf90222d85","merge_sha":"adcccf31a23c7eedff01f672159a3f4b724559b3"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5369118469?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5369118469?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`wasm32`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5369077005?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/adcccf31a23c7eedff01f672159a3f4b724559b3/checks/5369077005/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

nagisa commented:
@bors r+ spurious network failure?

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #94477

bors commented:
:pushpin: Commit df701a292ce552fddad2048cfcb6edaf90222d85 has been approved by `nagisa`

<!-- @bors r=nagisa df701a292ce552fddad2048cfcb6edaf90222d85 -->
<!-- homu: {"type":"Approved","sha":"df701a292ce552fddad2048cfcb6edaf90222d85","approver":"nagisa"} -->

nagisa commented:
@bors r-

nagisa commented:
@bors r=estebank

bors commented:
:pushpin: Commit df701a292ce552fddad2048cfcb6edaf90222d85 has been approved by `estebank`

<!-- @bors r=estebank df701a292ce552fddad2048cfcb6edaf90222d85 -->
<!-- homu: {"type":"Approved","sha":"df701a292ce552fddad2048cfcb6edaf90222d85","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit df701a292ce552fddad2048cfcb6edaf90222d85 with merge 7d6dc0530f5c4fcb230282f143f7ac71bedf105c...
<!-- homu: {"type":"BuildStarted","head_sha":"df701a292ce552fddad2048cfcb6edaf90222d85","merge_sha":"7d6dc0530f5c4fcb230282f143f7ac71bedf105c"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5379952881?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/7d6dc0530f5c4fcb230282f143f7ac71bedf105c/checks/5379952881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

matthiaskrgr commented:
@bors retry timeout

bors commented:
:hourglass: Testing commit df701a292ce552fddad2048cfcb6edaf90222d85 with merge 39a3b527674c1c8d2b9d3edc0cfae67abe6b3ecb...
<!-- homu: {"type":"BuildStarted","head_sha":"df701a292ce552fddad2048cfcb6edaf90222d85","merge_sha":"39a3b527674c1c8d2b9d3edc0cfae67abe6b3ecb"} -->

rust-timer commented:
Finished benchmarking commit (39a3b527674c1c8d2b9d3edc0cfae67abe6b3ecb): [comparison url](https://perf.rust-lang.org/compare.html?start=f0c4da49983aa699f715caf681e3154b445fb60b&end=39a3b527674c1c8d2b9d3edc0cfae67abe6b3ecb).

**Summary**: This benchmark run did not return any relevant results.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
